src/analysis/bounds.rs

```diff
@@ -380,34 +341,60 @@ mod tests {
-        assert_eq!(bounds.add_parameter("h", "", typ.clone(), false), false);
-        assert_eq!(bounds.add_parameter("i", "", typ.clone(), false), true);
-        assert_eq!(bounds.add_parameter("j", "", typ.clone(), false), true);
-        assert_eq!(bounds.add_parameter("k", "", typ.clone(), false), true);
-        assert_eq!(bounds.add_parameter("l", "", typ, false), false);
+        bounds.add_parameter("a", "", typ.clone(), false);
+        assert_eq!(bounds.iter().len(), 1);
+        bounds.add_parameter("b", "", typ.clone(), false);
+        bounds.add_parameter("c", "", typ.clone(), false);
+        bounds.add_parameter("d", "", typ.clone(), false);
+        bounds.add_parameter("e", "", typ.clone(), false);
+        bounds.add_parameter("f", "", typ.clone(), false);
+        bounds.add_parameter("g", "", typ.clone(), false);
+        bounds.add_parameter("h", "", typ.clone(), false);
+        assert_eq!(bounds.iter().len(), 8);
+        bounds.add_parameter("h", "", typ.clone(), false);
+        assert_eq!(bounds.iter().len(), 8);
+        bounds.add_parameter("i", "", typ.clone(), false);
+        bounds.add_parameter("j", "", typ.clone(), false);
+        bounds.add_parameter("k", "", typ, false);
+    }
+
+    #[test]
+    #[should_panic(expected = "No free type aliases!")]
+    fn exhaust_type_parameters() {
+        let mut bounds: Bounds = Default::default();
+        let typ = BoundType::NoWrapper;
+        for c in 'a'..='l' {
+            // Should panic on `l` because all type parameters are exhausted
+            bounds.add_parameter(c.to_string().as_str(), "", typ.clone(), false);
+        }
+    }
+
+    #[test]
+    fn get_parameter_bound() {
+        let mut bounds: Bounds = Default::default();
+        let typ = BoundType::NoWrapper;
+        bounds.add_parameter("a", "", typ.clone(), false);
+        bounds.add_parameter("b", "", typ.clone(), false);
+        let bound = bounds.get_parameter_bound("a").unwrap();
+        assert_eq!(bound.alias, Some('P'));
+        assert_eq!(bound.bound_type, typ);
+        let bound = bounds.get_parameter_bound("b").unwrap();
+        assert_eq!(bound.alias, Some('Q'));
+        assert_eq!(bound.bound_type, typ);
+        assert_eq!(bounds.get_parameter_bound("c"), None);
     }
 
     #[test]
-    fn get_parameter_alias_info() {
+    fn impl_bound() {
         let mut bounds: Bounds = Default::default();
         let typ = BoundType::IsA(None);
         bounds.add_parameter("a", "", typ.clone(), false);
         bounds.add_parameter("b", "", typ.clone(), false);
-        assert_eq!(
-            bounds.get_parameter_alias_info("a"),
-            Some(('P', typ.clone()))
-        );
-        assert_eq!(bounds.get_parameter_alias_info("b"), Some(('Q', typ)));
-        assert_eq!(bounds.get_parameter_alias_info("c"), None);
+        let bound = bounds.get_parameter_bound("a").unwrap();
+        assert_eq!(bound.alias, None);
+        assert_eq!(bound.bound_type, typ);
```

Comment:
Maybe we could have the same test for `BoundType::AsRef<None>`?

Reply:
I'm thinking of extending this with a few comments explaining why/when we do or do not expect a type alias.

In the end it shouldn't be based on `BoundType` as mentioned in a comment above, but I haven't invested the time to figure that one out and refactor it yet. Works well enough as a heuristic for now, but these tests should likely be adjusted in the future.